Remove AttributeList from Device Layer

ksperling-apple opened the pull request:
#### Summary

It was only used in a single place and probably didn't belong in the device layer in the first place.  Use a Span instead for DeviceInforProvider::SetUserLabelList.

Also simplify DeviceInforProvider::ClearUserLabelList().

#### Testing

Code is covered by existing tests.

ksperling-apple committed:
Remove AttributeList from Device Layer

It was only used in a single place and probably didn't belong in the device layer
in the first place.  Use a Span instead for DeviceInforProvider::SetUserLabelList.

Also simplify DeviceInforProvider::ClearUserLabelList().

ksperling-apple committed:
Add missing const to Span type

ksperling-apple commented:
/gemini review

andy31415 commented on src/app/clusters/user-label-server/user-label-cluster.cpp:
```diff
@@ -96,20 +85,22 @@ CHIP_ERROR WriteLabelList(const ConcreteDataAttributePath & path, AttributeValue
         {
-            ReturnErrorOnFailure(labelList.add(iter.GetValue()));
+            auto & label = iter.GetValue();
+            VerifyOrReturnError(IsValidLabelEntry(label), CHIP_IM_GLOBAL_STATUS(ConstraintError));
```
User label cluster is a code driven clusters, so it has testing infrastructure for unit tests.

Could we test this? Codecov says 58.33333% line coverage.

ksperling-apple committed:
Move max size constants into the header and add more tests

andy31415 committed:
Merge branch 'master' into platform-attributelist

ksperling-apple committed:
Allow the global DeviceInfoProvider to be reset to null

This is primarily useful for testing. Also remove the comment stating that
callers should assume the provider is not null; it starts out as null, and the
cluster code interacting with it already handles null gracefully.
